Translation/config refactor

dgonyeo opened the pull request:
This PR implements the refactor described in https://github.com/coreos/bugs/issues/2346.

The history is fairly granular right now, if it's too detailed I can squash related commits together.

Fixes https://github.com/coreos/bugs/issues/2346

ajeddeloh commented:
Before I start reviewing this, can you confirm the commit order GH is showing is correct and if it's not rebase with `--ignore-date`

dgonyeo commented:
This was rebased with `--ignore-date` I don't know what the hell github is doing on the `Conversation` tab, but the commit ordering on the `Commits` tab looks correct to me.

arithx commented on internal/config/types/maxversion.go:
```diff
@@ -0,0 +1,9 @@
+package types
```
apache header

dgonyeo commented on config/v1/config.go:
```diff
@@ -15,9 +15,9 @@
-	"errors"
 	"reflect"
 
+	"github.com/coreos/ignition/config/errors"
```
Combining the commit that adds the errors with the various commits that switch over to using them would make ordering tricky, since the commit that switches over `config/v1` to the errors depends on the work that makes `config/v1` use `config/util` for error handling, and that logic in `config/util` needs those errors.

I could squash together the commits across all the `config/v*` packages that switch over to using the new errors, if you think that would make this cleaner.

arithx replied:
Personally I think it's fine either way (leaving it or changing as you suggested), unless someone else has any other opinions on this.

dgonyeo replied:
I'll leave it then

dgonyeo commented on config/v2_0/translate.go:
```diff
@@ -1,25 +1,48 @@
+	"github.com/vincent-petithory/dataurl"
+)
+
+func intToPtr(x int) *int {
```
Yup, it would. Added a commit that does this.

dgonyeo commented on config/v2_0/translate.go:
```diff
@@ -1,25 +1,48 @@
 	config := types.Config{
 		Ignition: types.Ignition{
-			Version: v2_0.MaxVersion.String(),
+			Version: ConfigVersion,
```
`ConfigVersion` I actually intended to remove, it's redundant with the already-existing `types.MaxVersion`. I'll remove it.

Ideally this will only use things that already exist. Is there anything else in this commit (or others) that jumps out at you as using things from the future?

dgonyeo commented:
I've added changes to the `test` script that ensure that every go file in `config`, `validate`, `internal`, and `tests` have the Apache License header. There were a few files in the repo missing it, so I've also added it to them.

dgonyeo commented:
I just updated this with more commit body messages, which will hopefully help reviewers make sense of this.

dgonyeo commented on internal/exec/stages/files/files.go:
```diff
@@ -234,7 +234,15 @@ func (lst ByDirectorySegments) Swap(i, j int) {
+	return depth(lst[i].Node) < depth(lst[j].Node)
+}
+
+func depth(n types.Node) int {
```
The reason I made this an `int` is because [the function this is copied from](https://github.com/coreos/ignition/blob/eb48ac9a69ce2c3fe97d9b9bea5aae1c252d6911/config/types/node.go#L51) has this as an `int`.

If this ever gets large enough to reach the limits of an `int` we've probably got larger problems, but `uint` does make more sense for this function. I'll switch it to use `uint`.

arithx commented on doc/development.md:
````diff
@@ -10,20 +10,26 @@ sudo apt-get install libblkid-dev
 ./generate
 ```
 
+Add whatever validation logic is necessary to `config/v${LATEST_EXPERIMENTAL}/types`, modify the translator at `internal/config/translate.go` to handle the changes, and update `internal/config/translate_test.go` to properly test the changes.
````
Should also add a note that the blackbox tests will need an update in `tests/positive/general/general.go` for the VersionOnly set of tests.

dgonyeo replied:
Good catch. Fixed.

dgonyeo commented on config/v1/config.go:
```diff
@@ -32,24 +33,34 @@ var (
+		err = ErrVersion
+	}
+
+	rpt := validate.ValidateWithoutSource(reflect.ValueOf(config))
```
I still have to fix up the v1 validation. I'll make sure it happens before this gets merged in.

ajeddeloh commented:
This passes kola for qemu

ajeddeloh commented on config/util/helpers.go:
```diff
@@ -0,0 +1,30 @@
+// Copyright 2016 CoreOS, Inc.
```
2018

ajeddeloh commented on config/util/helpers.go:
```diff
@@ -0,0 +1,30 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
```
this can also replace internal/util/util.go.

It's StringToPtr is slightly different (doesn't have the `s == ""` check, but the only place it's used (in engine.go) it won't matter.

ajeddeloh replied:
Missed one in internal/config/translate.go

ajeddeloh commented on internal/exec/stages/files/files.go:
```diff
@@ -234,7 +234,15 @@ func (lst ByDirectorySegments) Swap(i, j int) {
+	return depth(lst[i].Node) < depth(lst[j].Node)
+}
+
+func depth(n types.Node) uint {
```
maybe make Depth() take a string, move it to config/util, and give it a name like PathDepth or something?

dgonyeo replied:
I'm not sure that makes sense because this is the only consumer of `depth()`. It's really just a helper function for `Less()` a few lines up and nothing more.

ajeddeloh replied:
Oh, sure then. I thought there was some validation that used it. /shrug

dgonyeo commented:
Just added a commit that updates the `config/v1/types` validation logic to use the same validation pattern as everything else. We now get line/col information for v1 validation!

```
derek@proton ~/go/src/github.com/coreos/ignition> cat config.ign 
{
  "ignitionVersion": 1,
  "storage": {
    "filesystems": [
      {
        "device": "this/is/a/relative/path",
        "format": "ext4"
      }
    ]
  }
}
derek@proton ~/go/src/github.com/coreos/ignition> ./bin/amd64/ignition-validate config.ign
error at line 6, column 44
    5:       {
    6:         "device": "this/is/a/relative/path"
                                                 ^
path not absolute
```

Just what our users always wanted.

dgonyeo commented:
retest this please

ajeddeloh commented on config/v1/types/filesystem_test.go:
```diff
@@ -77,118 +43,14 @@ func TestFilesystemFormatAssertValid(t *testing.T) {
-		},
-		{
-			in:  in{data: `{"format": "ext4"}`},
-			out: out{filesystem: Filesystem{Format: "ext4"}, err: ErrPathRelative},
```
This is being dropped since it's covered by the path valiation?

dgonyeo replied:
Yup

ajeddeloh commented on config/v1/types/filesystem_test.go:
```diff
@@ -77,118 +43,14 @@ func TestFilesystemFormatAssertValid(t *testing.T) {
-		out out
-	}{
-		{
-			in:  in{filesystem: Filesystem{Device: "/foo", Format: "ext4"}},
```
Looks like we're losing these tests.

dgonyeo replied:
There is no longer an `AssertValid` for `Filesystem` since it just called `AssertValid` on two of its fields of types `FilesystemFormat` and `Path`. The logic that finds and calls the `Validate` functions is what guarantees that the sub-`Validate`s are called, and there are separate tests for `Validate` on both of those fields ([here](https://github.com/coreos/ignition/pull/519/files/530602f5a9eae4b25e85a425befc49ec99b02cba..c667b7faec249bfe32db53595a4b843cfd26a66f#diff-b5055fb9d5075288ed7b1b69531a5e2fR24) and [here](https://github.com/coreos/ignition/pull/519/files/530602f5a9eae4b25e85a425befc49ec99b02cba..c667b7faec249bfe32db53595a4b843cfd26a66f#diff-adfe4224d4d2d5a075e2183a18298cdaR24)).
